src/translator/batch_translator.cpp

```diff
@@ -22,17 +22,17 @@ void BatchTranslator::initialize() {
   bool check = options_->get<bool>("check-bytearray",false); // Flag holds whether validate the bytearray (model and shortlist)
   if (options_->hasAndNotEmpty("shortlist")) {
     int srcIdx = 0, trgIdx = 1;
-    bool shared_vcb = vocabs_->front() == vocabs_->back();
+    bool shared_vcb = vocabs_->source().front() == vocabs_->target();
     if (shortlistMemory_->size() > 0 && shortlistMemory_->begin() != nullptr) {
       slgen_ = New<data::BinaryShortlistGenerator>(shortlistMemory_->begin(), shortlistMemory_->size(),
-                                                     vocabs_->front(), vocabs_->back(),
+                                                   vocabs_->source().front(), vocabs_->target(),
                                                      srcIdx, trgIdx, shared_vcb, check);
     }
     else {
       // Changed to BinaryShortlistGenerator to enable loading binary shortlist file
       // This class also supports text shortlist file
-      slgen_ = New<data::BinaryShortlistGenerator>(options_, vocabs_->front(),
-                                                    vocabs_->back(), srcIdx,
+      slgen_ = New<data::BinaryShortlistGenerator>(options_, vocabs_->source().front(),
+                                                    vocabs_->target(), srcIdx,
```

Comment:
@XapaJIaMnu Can you shed some light here? `ShortlistGenerator` seem to be operating  with only one source-vocab. What becomes in the ensembling case? Is there an advantage to providing an alias through a member function to only the first source?

Reply:
I added comments for why we invoke `vocabs_->source().front()`(because we currently only use one source vocab).

Reply:
I'd want to hide away the `.front()` tail (access skipping `Vocabs` layer) by an alias (`vocabs_.source()` or `vocabs_.source_first()`) which gives just the first vocab (so it's easier to edit later within the `Vocab` class). This skipped access violates the attempt to freely vary internals of `Vocabs` while outside wiring remains functional. Hence asking Nick or someone to confirm how `ShortListGenerator` works with ensembling.

Reply:
As you can see from the beamsearch, each scorer could have its own shortlist. https://github.com/marian-nmt/marian-dev/blob/fe74576dc3e957f97b4842951fb6ee7ec52fa324/src/translator/beam_search.cpp#L434

However in practise when decoding with ensembles you don't use a shortlist.

Don't do `source_first()` method. Just don't. Don't use such aliases please, they are not making the code more readable or simpler to understand.

Reply:
That clarifies enough not to keep a first on the shortlist. 

@qianqianzhu we're looking at a vector like init on shortlist similar to the model and zipping both to set respective shortlist on each scorer, looking at batch translator, eventually? That will get rid of the first() in BatchTranslator. 

As for preprocessing, next place with shortlist, I think I'd probably have to rework a lot I guess. 

Marking this as resolved for purposes of this PR.